File: backend/FinanceTracker.Application/Common/MonthKey.cs
using System.Globalization;

namespace FinanceTracker.Application.Common;

/// <summary>Helpers for the <c>YYYY-MM</c> month keys used by budgets, dashboards and reports.</summary>
public static class MonthKey
{
    public const string Format = "yyyy-MM";

    public static string From(DateTime value) => value.ToString(Format, CultureInfo.InvariantCulture);

    public static bool TryParse(string? value, out DateTime firstDayUtc)
    {
        firstDayUtc = default;
        if (string.IsNullOrWhiteSpace(value)
            || !DateTime.TryParseExact(value, Format, CultureInfo.InvariantCulture, DateTimeStyles.None, out var parsed))
        {
            return false;
        }

        firstDayUtc = FirstDayUtc(parsed.Year, parsed.Month);
        return true;
    }

    /// <summary>Parses a month key or throws a 400.</summary>
    public static DateTime Parse(string? value) =>
        TryParse(value, out var start)
            ? start
            : throw AppException.BadRequest("Month must be in YYYY-MM format.");

    /// <summary>The first day of the month containing <paramref name="value"/>, in UTC.</summary>
    public static DateTime StartOfMonth(DateTime value) => FirstDayUtc(value.Year, value.Month);

    /// <summary>Midnight UTC on the first day of the given month.</summary>
    public static DateTime FirstDayUtc(int year, int month) =>
        new(year, month, 1, 0, 0, 0, DateTimeKind.Utc);

    /// <summary>The <paramref name="count"/> months ending with the month of <paramref name="reference"/>, oldest first.</summary>
    public static IReadOnlyList<DateTime> TrailingMonths(DateTime reference, int count)
    {
        var last = StartOfMonth(reference);
        return Enumerable.Range(0, count).Select(i => last.AddMonths(i - count + 1)).ToList();
    }
}

File: backend-go/internal/application/monthkey.go
package application

import (
	"time"

	"github.com/pharaujo/finance/backend-go/internal/domain"
)

// MonthFormat is the layout of the YYYY-MM keys budgets, dashboards and reports
// use. It is the Go spelling of MonthKey.Format ("yyyy-MM").
const MonthFormat = "2006-01"

// monthFormatMessage is the 400 MonthKey.Parse raises, byte for byte.
const monthFormatMessage = "Month must be in YYYY-MM format."

// MonthFrom renders the month key of a timestamp.
func MonthFrom(value time.Time) string { return value.UTC().Format(MonthFormat) }

// TryParseMonth returns midnight UTC on the first day of the named month.
func TryParseMonth(value string) (time.Time, bool) {
	parsed, err := time.ParseInLocation(MonthFormat, value, time.UTC)
	if err != nil {
		return time.Time{}, false
	}
	return FirstDayUTC(parsed.Year(), parsed.Month()), true
}

// ParseMonth is TryParseMonth or a 400.
func ParseMonth(value string) (time.Time, error) {
	start, ok := TryParseMonth(value)
	if !ok {
		return time.Time{}, domain.BadRequest(monthFormatMessage)
	}
	return start, nil
}

// StartOfMonth is the first day of the month containing value, in UTC.
func StartOfMonth(value time.Time) time.Time {
	utc := value.UTC()
	return FirstDayUTC(utc.Year(), utc.Month())
}

// FirstDayUTC is midnight UTC on the first day of the given month.
func FirstDayUTC(year int, month time.Month) time.Time {
	return time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
}

// TrailingMonths returns the count months ending with the month of reference,
// oldest first.
func TrailingMonths(reference time.Time, count int) []time.Time {
	last := StartOfMonth(reference)
	months := make([]time.Time, 0, count)
	for i := range count {
		months = append(months, domain.AddMonths(last, i-count+1))
	}
	return months
}
